In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
import seaborn as sns
sns.set()

from scipy import stats
stats.chisqprob = lambda chisq, df:stats.chi12.sf(chisq,df)

In [3]:
raw_data = pd.read_csv('framingham_heart_disease 2.csv')
raw_data.describe(include = 'all')

male          age    education  currentSmoker   cigsPerDay  \
count  4238.000000  4238.000000  4133.000000    4238.000000  4209.000000   
mean      0.429212    49.584946     1.978950       0.494101     9.003089   
std       0.495022     8.572160     1.019791       0.500024    11.920094   
min       0.000000    32.000000     1.000000       0.000000     0.000000   
25%       0.000000    42.000000     1.000000       0.000000     0.000000   
50%       0.000000    49.000000     2.000000       0.000000     0.000000   
75%       1.000000    56.000000     3.000000       1.000000    20.000000   
max       1.000000    70.000000     4.000000       1.000000    70.000000   

            BPMeds  prevalentStroke  prevalentHyp     diabetes      totChol  \
count  4185.000000      4238.000000   4238.000000  4238.000000  4188.000000   
mean      0.029630         0.005899      0.310524     0.025720   236.721585   
std       0.169584         0.076587      0.462763     0.158316    44.590334   
min       0.000000         0.000000      0.000000     0.000000   107.000000   
25%       0.000000         0.000000      0.000000     0.000000   206.000000   
50%       0.000000         0.000000      0.000000     0.000000   234.000000   
75%       0.000000         0.000000      1.000000     0.000000   263.000000   
max       1.000000         1.000000      1.000000     1.000000   696.000000   

             sysBP        diaBP          BMI    heartRate      glucose  \
count  4238.000000  4238.000000  4219.000000  4237.000000  3850.000000   
mean    132.352407    82.893464    25.802008    75.878924    81.966753   
std      22.038097    11.910850     4.080111    12.026596    23.959998   
min      83.500000    48.000000    15.540000    44.000000    40.000000   
25%     117.000000    75.000000    23.070000    68.000000    71.000000   
50%     128.000000    82.000000    25.400000    75.000000    78.000000   
75%     144.000000    89.875000    28.040000    83.000000    87.000000   
max     295.000000   142.500000    56.800000   143.000000   394.000000   

        TenYearCHD  
count  4238.000000  
mean      0.151958  
std       0.359023  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       0.000000  
max       1.000000

In [4]:
raw_data.isnull().sum()

male                 0
age                  0
education          105
currentSmoker        0
cigsPerDay          29
BPMeds              53
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             50
sysBP                0
diaBP                0
BMI                 19
heartRate            1
glucose            388
TenYearCHD           0
dtype: int64

In [5]:
data_no_mv = raw_data.dropna(axis=0)

In [6]:
data_no_mv.isnull().sum()

male               0
age                0
education          0
currentSmoker      0
cigsPerDay         0
BPMeds             0
prevalentStroke    0
prevalentHyp       0
diabetes           0
totChol            0
sysBP              0
diaBP              0
BMI                0
heartRate          0
glucose            0
TenYearCHD         0
dtype: int64

In [7]:
data = data_no_mv.drop(['education'], axis = 1)
data

male  age  currentSmoker  cigsPerDay  BPMeds  prevalentStroke  \
0        1   39              0         0.0     0.0                0   
1        0   46              0         0.0     0.0                0   
2        1   48              1        20.0     0.0                0   
3        0   61              1        30.0     0.0                0   
4        0   46              1        23.0     0.0                0   
...    ...  ...            ...         ...     ...              ...   
4231     1   58              0         0.0     0.0                0   
4232     1   68              0         0.0     0.0                0   
4233     1   50              1         1.0     0.0                0   
4234     1   51              1        43.0     0.0                0   
4237     0   52              0         0.0     0.0                0   

      prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  heartRate  \
0                0         0    195.0  106.0   70.0  26.97       80.0   
1                0         0    250.0  121.0   81.0  28.73       95.0   
2                0         0    245.0  127.5   80.0  25.34       75.0   
3                1         0    225.0  150.0   95.0  28.58       65.0   
4                0         0    285.0  130.0   84.0  23.10       85.0   
...            ...       ...      ...    ...    ...    ...        ...   
4231             1         0    187.0  141.0   81.0  24.96       80.0   
4232             1         0    176.0  168.0   97.0  23.14       60.0   
4233             1         0    313.0  179.0   92.0  25.97       66.0   
4234             0         0    207.0  126.5   80.0  19.71       65.0   
4237             0         0    269.0  133.5   83.0  21.47       80.0   

      glucose  TenYearCHD  
0        77.0           0  
1        76.0           0  
2        70.0           0  
3       103.0           1  
4        85.0           0  
...       ...         ...  
4231     81.0           0  
4232     79.0           1  
4233     86.0           1  
4234     68.0           0  
4237    107.0           0  

[3656 rows x 15 columns]

## Defining Features and Targets

In [8]:
y = data['TenYearCHD']
x1 = data.drop(['TenYearCHD'], axis = 1)

## Creating the Regression

In [9]:
x = sm.add_constant(x1)
reg_log = sm.Logit(y,x)
results_log = reg_log.fit()
results_log.summary()

Optimization terminated successfully.
         Current function value: 0.376795
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:             TenYearCHD   No. Observations:                 3656
Model:                          Logit   Df Residuals:                     3641
Method:                           MLE   Df Model:                           14
Date:                Thu, 05 Mar 2020   Pseudo R-squ.:                  0.1171
Time:                        18:51:08   Log-Likelihood:                -1377.6
converged:                       True   LL-Null:                       -1560.3
Covariance Type:            nonrobust   LLR p-value:                 2.412e-69
===================================================================================
                      coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
const              -8.4895      0.695    -12.221      0.000      -9.851      -7.128
male                0.5540      0.109      5.082      0.000       0.340       0.768
age                 0.0642      0.007      9.676      0.000       0.051       0.077
currentSmoker       0.0713      0.157      0.455      0.649      -0.236       0.378
cigsPerDay          0.0180      0.006      2.887      0.004       0.006       0.030
BPMeds              0.1576      0.234      0.673      0.501      -0.301       0.616
prevalentStroke     0.7060      0.489      1.444      0.149      -0.253       1.665
prevalentHyp        0.2332      0.138      1.689      0.091      -0.037       0.504
diabetes            0.0430      0.315      0.136      0.892      -0.575       0.661
totChol             0.0023      0.001      2.009      0.045    5.53e-05       0.004
sysBP               0.0157      0.004      4.139      0.000       0.008       0.023
diaBP              -0.0046      0.006     -0.718      0.473      -0.017       0.008
BMI                 0.0081      0.013      0.635      0.525      -0.017       0.033
heartRate          -0.0031      0.004     -0.731      0.465      -0.011       0.005
glucose             0.0071      0.002      3.185      0.001       0.003       0.012
===================================================================================
"""

In [10]:
data_cleaned = x.copy()
data_cleaned = data_cleaned.drop(['BPMeds'], axis = 1)

In [11]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train, y_test = train_test_split(data_cleaned, y, test_size=0.2, random_state = 365)

In [12]:
x_cleaned = x_train
x_new = sm.add_constant(x_cleaned)
reg_log = sm.Logit(y_train,x_new)
results_log = reg_log.fit()
results_log.summary()

Optimization terminated successfully.
         Current function value: 0.384151
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:             TenYearCHD   No. Observations:                 2924
Model:                          Logit   Df Residuals:                     2910
Method:                           MLE   Df Model:                           13
Date:                Thu, 05 Mar 2020   Pseudo R-squ.:                  0.1114
Time:                        18:51:10   Log-Likelihood:                -1123.3
converged:                       True   LL-Null:                       -1264.1
Covariance Type:            nonrobust   LLR p-value:                 1.631e-52
===================================================================================
                      coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
const              -8.0750      0.757    -10.665      0.000      -9.559      -6.591
male                0.4672      0.120      3.887      0.000       0.232       0.703
age                 0.0648      0.007      8.782      0.000       0.050       0.079
currentSmoker       0.0074      0.175      0.042      0.966      -0.335       0.349
cigsPerDay          0.0205      0.007      2.954      0.003       0.007       0.034
prevalentStroke     0.7210      0.578      1.247      0.212      -0.412       1.854
prevalentHyp        0.2138      0.152      1.407      0.159      -0.084       0.512
diabetes            0.0407      0.354      0.115      0.909      -0.654       0.735
totChol             0.0012      0.001      1.001      0.317      -0.001       0.004
sysBP               0.0161      0.004      3.829      0.000       0.008       0.024
diaBP              -0.0088      0.007     -1.226      0.220      -0.023       0.005
BMI                 0.0136      0.014      0.967      0.334      -0.014       0.041
heartRate          -0.0025      0.005     -0.537      0.591      -0.012       0.007
glucose             0.0073      0.002      3.011      0.003       0.003       0.012
===================================================================================
"""

In [13]:
def confusion_matrix(data,actual_values,model):
        
        # Confusion matrix 
        
        # Parameters
        # ----------
        # data: data frame or array
            # data is a data frame formatted in the same way as your input data (without the actual values)
            # e.g. const, var1, var2, etc. Order is very important!
        # actual_values: data frame or array
            # These are the actual values from the test_data
            # In the case of a logistic regression, it should be a single column with 0s and 1s
            
        # model: a LogitResults object
            # this is the variable where you have the fitted model 
            # e.g. results_log in this course
        # ----------

        #Predict the values using the Logit model
        pred_values = model.predict(data)
        # Specify the bins 
        bins=np.array([0,0.5,1])
        # Create a histogram, where if values are between 0 and 0.5 tell will be considered 0
        # if they are between 0.5 and 1, they will be considered 1
        cm = np.histogram2d(actual_values, pred_values, bins=bins)[0]
        # Calculate the accuracy
        accuracy = (cm[0,0]+cm[1,1])/cm.sum()
        # Return the confusion matrix and
        string = 'Accuracy is ' + repr(accuracy*100)+' %'
        return cm, string

In [14]:
confusion_matrix(x_new,y_train,results_log)

(array([[2455.,   14.],
        [ 414.,   41.]]),
 'Accuracy is 85.3625170998632 %')

## Testing Model

In [15]:
x_test1 = x_test
reg_log = sm.Logit(y_test,x_test)
results_log1 = reg_log.fit()
results_log1.summary()

Optimization terminated successfully.
         Current function value: 0.337696
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:             TenYearCHD   No. Observations:                  732
Model:                          Logit   Df Residuals:                      718
Method:                           MLE   Df Model:                           13
Date:                Thu, 05 Mar 2020   Pseudo R-squ.:                  0.1636
Time:                        18:51:11   Log-Likelihood:                -247.19
converged:                       True   LL-Null:                       -295.56
Covariance Type:            nonrobust   LLR p-value:                 7.103e-15
===================================================================================
                      coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
const             -10.9762      1.820     -6.030      0.000     -14.544      -7.409
male                0.9580      0.265      3.610      0.000       0.438       1.478
age                 0.0637      0.016      4.094      0.000       0.033       0.094
currentSmoker       0.2822      0.369      0.765      0.444      -0.440       1.005
cigsPerDay          0.0086      0.015      0.587      0.557      -0.020       0.037
prevalentStroke     0.9468      0.895      1.057      0.290      -0.808       2.702
prevalentHyp        0.3507      0.331      1.058      0.290      -0.299       1.000
diabetes            0.1723      0.708      0.243      0.808      -1.216       1.561
totChol             0.0078      0.003      2.778      0.005       0.002       0.013
sysBP               0.0170      0.009      1.920      0.055      -0.000       0.034
diaBP               0.0111      0.015      0.754      0.451      -0.018       0.040
BMI                -0.0126      0.030     -0.416      0.677      -0.072       0.047
heartRate          -0.0048      0.010     -0.468      0.640      -0.025       0.015
glucose             0.0053      0.006      0.872      0.383      -0.007       0.017
===================================================================================
"""

In [16]:
def confusion_matrix(data,actual_values,model):
        
        # Confusion matrix 
        
        # Parameters
        # ----------
        # data: data frame or array
            # data is a data frame formatted in the same way as your input data (without the actual values)
            # e.g. const, var1, var2, etc. Order is very important!
        # actual_values: data frame or array
            # These are the actual values from the test_data
            # In the case of a logistic regression, it should be a single column with 0s and 1s
            
        # model: a LogitResults object
            # this is the variable where you have the fitted model 
            # e.g. results_log in this course
        # ----------

        #Predict the values using the Logit model
        pred_values = model.predict(data)
        # Specify the bins 
        bins=np.array([0,0.5,1])
        # Create a histogram, where if values are between 0 and 0.5 tell will be considered 0
        # if they are between 0.5 and 1, they will be considered 1
        cm = np.histogram2d(actual_values, pred_values, bins=bins)[0]
        # Calculate the accuracy
        accuracy = (cm[0,0]+cm[1,1])/cm.sum()
        # Return the confusion matrix and
        string = 'Accuracy is ' + repr(accuracy*100)+' %'
        return cm, string

In [17]:
confusion_matrix(x_new,y_train,results_log)

(array([[2455.,   14.],
        [ 414.,   41.]]),
 'Accuracy is 85.3625170998632 %')

In [18]:
cm_df = pd.DataFrame(results_log.pred_table())
cm_df.columns = ['Predicted 0','Predicted 1']
cm_df = cm_df.rename(index={0:'Actual 0',1:'Actual 1'})
cm_df

Predicted 0  Predicted 1
Actual 0       2455.0         14.0
Actual 1        414.0         41.0

## Hence, the model produced an accurate prediction 2496 times out of 2924 times resulting in,
# Accuracy = 85.36%